src/expr/src/vector_op/cmp.rs

```diff
@@ -117,48 +118,58 @@ where
         (None, None) => Ok(Some(false)),
     }
 }
-
-#[inline(always)]
-fn str_cmp<F>(l: &str, r: &str, func: F) -> Result<bool>
-where
-    F: FnOnce(&str, &str) -> bool,
-{
-    Ok(func(l, r))
-}
-
-#[inline(always)]
-pub fn str_eq(l: &str, r: &str) -> Result<bool> {
-    str_cmp(l, r, |a, b| a == b)
+#[derive(Clone, Copy, Debug)]
+pub enum Operation {
+    Eq,
+    Ne,
+    Lt,
+    Gt,
+    Le,
+    Ge,
 }
 
-#[inline(always)]
-pub fn str_ne(l: &str, r: &str) -> Result<bool> {
-    str_cmp(l, r, |a, b| a != b)
-}
+pub(crate) static EQ: Operation = Operation::Eq;
+pub(crate) static NE: Operation = Operation::Ne;
+pub(crate) static LT: Operation = Operation::Lt;
+pub(crate) static GT: Operation = Operation::Gt;
+pub(crate) static LE: Operation = Operation::Le;
+pub(crate) static GE: Operation = Operation::Ge;
 
 #[inline(always)]
-pub fn str_ge(l: &str, r: &str) -> Result<bool> {
-    str_cmp(l, r, |a, b| a >= b)
+pub fn struct_cmp(op: Operation) -> fn(StructRef, StructRef) -> Result<bool> {
+    use crate::gen_cmp;
+    gen_cmp!(op)
 }
 
 #[inline(always)]
-pub fn str_gt(l: &str, r: &str) -> Result<bool> {
-    str_cmp(l, r, |a, b| a > b)
+pub fn list_cmp(op: Operation) -> fn(ListRef, ListRef) -> Result<bool> {
+    use crate::gen_cmp;
+    gen_cmp!(op)
 }
 
 #[inline(always)]
-pub fn str_le(l: &str, r: &str) -> Result<bool> {
-    str_cmp(l, r, |a, b| a <= b)
+pub fn str_cmp(op: Operation) -> fn(&str, &str) -> Result<bool> {
+    use crate::gen_cmp;
+    gen_cmp!(op)
 }
 
-#[inline(always)]
-pub fn str_lt(l: &str, r: &str) -> Result<bool> {
-    str_cmp(l, r, |a, b| a < b)
+#[macro_export]
+macro_rules! gen_cmp {
+    ($op:expr) => {
+        match $op {
+            Operation::Eq => |l, r| Ok(l == r),
+            Operation::Ne => |l, r| Ok(l != r),
+            Operation::Lt => |l, r| Ok(l < r),
+            Operation::Gt => |l, r| Ok(l > r),
+            Operation::Le => |l, r| Ok(l <= r),
+            Operation::Ge => |l, r| Ok(l >= r),
+        }
+    };
 }
 
 pub fn str_is_distinct_from(l: Option<&str>, r: Option<&str>) -> Result<Option<bool>> {
     match (l, r) {
-        (Some(lv), Some(rv)) => Ok(str_ne(lv, rv).ok()),
+        (Some(lv), Some(rv)) => Ok(str_cmp(Operation::Ne)(lv, rv).ok()),
```

Comment:
We shouldn't use enum dispatching at this function. It will lead to bad-optimized code.